lib/ansible/modules/windows/win_say.ps1

```diff
@@ -3,51 +3,61 @@
 # Copyright: (c) 2016, Jon Hawkesworth (@jhawkesworth) <figs@unity.demon.co.uk>
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
-#Requires -Module Ansible.ModuleUtils.Legacy
+#AnsibleRequires -CSharpUtil Ansible.Basic
+
+$spec = @{
+   options = @{
+      msg = @{ type = "str"  }
+      msg_file = @{ type = "path"  }
+      start_sound_path = @{ type = "path"  }
+      end_sound_path = @{ type = "path"  }
+      voice = @{ type = "str"  }
+      speech_speed = @{ type = "int"; default = 0  }
+   }
+   supports_check_mode = $true
+}
 
-$params = Parse-Args $args -supports_check_mode $true
-$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+$module = [Ansible.Basic.AnsibleModule]::Create($args, $spec)
 
-$msg = Get-AnsibleParam -obj $params -name "msg" -type "str"
-$msg_file = Get-AnsibleParam -obj $params -name "msg_file" -type "path"
-$start_sound_path = Get-AnsibleParam -obj $params -name "start_sound_path" -type "path"
-$end_sound_path = Get-AnsibleParam -obj $params -name "end_sound_path" -type "path"
-$voice = Get-AnsibleParam -obj $params -name "voice" -type "str"
-$speech_speed = Get-AnsibleParam -obj $params -name "speech_speed" -type "int" -default 0
 
-$result = @{
-    changed = $false
-}
+$msg = $module.Params.msg
+$msg_file = $module.Params.msg_file
+$start_sound_path = $module.Params.start_sound_path
+$end_sound_path = $module.Params.end_sound_path
+$voice = $module.Params.voice
+$speech_speed = $module.Params.speech_speed
+
+$module.Result.changed = $false
 
 $words = $null
 
 if ($speech_speed -lt -10 -or $speech_speed -gt 10) {
-   Fail-Json $result "speech_speed needs to a integer in the range -10 to 10.  The value $speech_speed is outside this range."
+   $module.FailJson("speech_speed needs to be an integer in the range -10 to 10.  The value $speech_speed is outside this range.")
 }
 
 if ($msg_file -and $msg) {
-   Fail-Json $result "Please specify either msg_file or msg parameters, not both"
+   $module.FailJson("Please specify either msg_file or msg parameters, not both")
```

Comment:
You can add to the spec the `mutually_exclusive` option which removes the need for this. In this example you would do

```
$spec = @{
    ...
    mutually_exclusive = @(
        ,@('msg', 'msg_file')
    )
}
```

The leading `,` in the mutually_exclusive list is required when you only have 1 mutually_exclusive pair. This is because PowerShell will just convert a single array inside another array to just one array. The `,` forces it to keep the inner array separately. If you have more mutually exclusive pairs then just add it as a new inner array with the mutually exclusive options as the values for those inner arrays.

The valid options for the spec can be found under https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development.

Reply:
Thanks, makes sense.  Will fix